concordium-std-derive/src/lib.rs

```diff
@@ -33,47 +37,275 @@ fn attach_error<A>(mut v: syn::Result<A>, msg: &str) -> syn::Result<A> {
+                            format!("'{}' also appears here.", ident),
+                        ));
+                    } else {
+                        ret.flags.insert(ident.clone());
+                    }
+                } else {
+                    errors.push(syn::Error::new(
+                        p.span(),
+                        "Unrecognized attribute. Only attribute names consisting of a single \
+                         identifier are recognized.",
+                    ))
                 }
             }
             Meta::List(p) => {
-                if p.path.is_ident(name) {
-                    return Err(syn::Error::new(
-                        attr.span(),
-                        format!("The `{}` attribute must have a string literal value.", name),
-                    ));
-                }
+                errors.push(syn::Error::new(p.span(), "Unrecognized attribute."));
             }
         }
     }
-    Ok(None)
+    // TODO: Replace with into_values when least rust version becomes 1.54.
+    let mut iter = errors
+        .into_iter()
+        .chain(duplicate_values.into_iter().map(|(_, v)| v))
+        .chain(duplicate_flags.into_iter().map(|(_, v)| v));
+    // If there are any errors we combine them.
+    if let Some(err) = iter.next() {
+        let mut err = err;
+        for next_err in iter {
+            err.combine(next_err);
+        }
+        Err(err)
+    } else {
+        Ok(ret)
+    }
+}
+
+#[cfg(feature = "build-schema")]
+/// Attributes applicable to the `contract_state` annotation.
+struct ContractStateAttributes {
+    /// Name of the contract the contract state applies to.
+    pub(crate) contract: syn::LitStr,
+}
+
+#[cfg(feature = "build-schema")]
+/// Parse nested attributes to the `contract_state` attribute.
+fn parse_contract_state_attributes<'a, I: IntoIterator<Item = &'a Meta>>(
+    attrs: I,
+) -> syn::Result<ContractStateAttributes> {
+    let mut attributes = parse_attributes(attrs)?;
+    let contract = attributes.extract_value("contract").ok_or_else(|| {
+        syn::Error::new(
+            Span::call_site(),
+            "A name for the contract must be provided, using the 'contract' attribute.\n\nFor \
+             example, #[contract_state(contract = \"my-contract\")]",
+        )
+    })?;
+    attributes.report_all_attributes()?;
+    Ok(ContractStateAttributes {
+        contract,
+    })
+}
+
+fn parse_init_attributes<'a, I: IntoIterator<Item = &'a Meta>>(
+    attrs: I,
+) -> syn::Result<InitAttributes> {
+    let mut attributes = parse_attributes(attrs)?;
+    let contract: syn::LitStr = attributes.extract_value("contract").ok_or_else(|| {
+        syn::Error::new(
+            Span::call_site(),
+            "A name for the contract must be provided, using the 'contract' attribute.\n\nFor \
+             example, #[init(contract = \"my-contract\")]",
+        )
+    })?;
+    let parameter: Option<syn::LitStr> = attributes.extract_value("parameter");
```

Comment:
Consider moving the string literals for attribute names (`"parameter"`, `"payable"`, `"enable_logger"` and so on) into `const`s.

Reply:
To be honest I don't really see the advantage of this. What is your reasoning?

Reply:
They appear in multiple places and moving them to a const would make the compiler help ensuring the spelling is the same.

Another minor advantage is that it is easier to ensure we cover every parameter, since they would be defined in the same place, and we would get warnings from unused const.

Reply:
I don't really agree with this reasoning but I have made this change since I want to get this merged.

In my view the code is now harder to read due to an indirection.